Add support for bf16/fp32 mix precision in cuda_batch_norm

robert1003 opened the pull request:
Fixes #ISSUE_NUMBER

* CUDNN's batch norm kernel supports bf16 input + fp32 weight input ([docs](https://docs.nvidia.com/deeplearning/cudnn/api/cudnn-ops-library.html#cudnnbatchnormalizationbackwardEx)) but pytorch does not. This PR allows it.
* Fix a bug in constant type. Constants (alpha and beta) should have the same type as the weights not the input.

robert1003 committed:
add support for bf16/fp32 mix precision in cuda_batch_norm

eqy reviewed (commented):
Should tests be added/updated to test this on GPUs >= sm80?

robert1003 commented:
Hi @eqy yes that make sense, could you let me know where is the test file I should modify and how should I run the test? Sorry I am not really familiar with the overall codebase...

robert1003 commented:
Got it thanks will do it this weekend!

robert1003 committed:
Update Normalization.cpp

Add check for device capability

robert1003 committed:
Update BatchNorm.cpp

alpha and beta should have the same type as weight not input

robert1003 committed:
fix linter error

robert1003 committed:
add test for batchnorm + bfloat16

robert1003 commented:
@eqy I added a test in `test/test_nn.py` to test batchnorm in cudnn + bf16 setting. I think the mixed_precision test is failing because of the bug I found, which I have fixed in subsequent commits. Could you take a look? Thanks!

robert1003 committed:
fix linter

robert1003 committed:
cache and simplify the check

robert1003 commented:
@eqy It seems that some of the errors in CIs are not caused by my changes:
```
/var/lib/jenkins/workspace/aten/src/ATen/cuda/CUDAContextLight.h:6:10: fatal error: cuda_runtime_api.h: No such file or directory
    6 | #include <cuda_runtime_api.h>
```

Is this expected? Thanks!

eqy commented:
Might need a rebase after resolving merge conflicts; you can do this with `@pytorchmergebot rebase`

robert1003 commented:
@pytorchmergebot rebase

robert1003 committed:
Merge branch 'main' of https://github.com/pytorch/pytorch

robert1003 commented:
I think the compile error is caused by trying to import `#include <ATen/cuda/CUDAContext.h>` for cpu machines. Is there a flag for coda such that I can use for `#ifdef <FLAG>`? 

Thank you!

robert1003 committed:
add compile if/else for CUDA

robert1003 committed:
fix compile error

robert1003 commented:
@eqy I'm not sure if `#ifdef USE_CUDA` is the correct switch to use...

eqy commented on aten/src/ATen/native/Normalization.cpp:
```diff
@@ -58,6 +58,10 @@
 #endif
 
+#ifdef USE_CUDA
+#include <ATen/cuda/CUDAContext.h>
```
My understanding is that `Normalization.cpp` shouldn't depend on `CUDAContext.h`; the typical way that this kind of support check is done is via `CUDAHooks.cpp` which can query things like the device properties if CUDA is available---`Normalization.cpp` can call into `CUDAHooks` without needing to depend on anything CUDA-specific. You can see examples of other support-checking functions https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/cuda/detail/CUDAHooks.cpp

robert1003 committed:
update version check

robert1003 committed:
add import file

robert1003 commented:
@eqy Is this PR ready to merge?

eqy commented on test/test_nn.py:
```diff
@@ -4897,6 +4897,26 @@ def test_batchnorm_cudnn_half(self):
+    @unittest.skipIf(not TEST_CUDA, "CUDA unavailable")
+    def test_batchnorm_cudnn_bf16(self):
+        # THNN
+        input = torch.randint(1, 10, (2, 3, 2, 2), dtype=torch.bfloat16, device="cuda", requires_grad=True)
```
How is this choosing which backend to dispatch to? Would we need a context manager or something else to prevent it from going to cuDNN to generate the reference?

eqy commented:
@robert1003 I think it looks mostly okay, see the one comment on how the reference is generated for the test

Otherwise please rebase so the failing tests can be rerun

robert1003 committed:
Merge branch 'pytorch:main' into main

robert1003 commented:
I just update my branch with the newest commit on PyTorch master!

@pytorchmergebot rebase

robert1003 commented:
@eqy I don't think I have the permission to use the rebase bot...

eqy commented:
@pytorchmergebot rebase
